dev/MRTCore/mrt/Microsoft.Windows.ApplicationModel.Resources/src/ResourceContext.cpp

```diff
@@ -1,15 +1,76 @@
-﻿// Copyright (c) Microsoft Corporation and Contributors.
+// Copyright (c) Microsoft Corporation and Contributors.
 // Licensed under the MIT License.
 
 #include "pch.h"
 #include "ResourceContext.h"
 #include "ResourceContext.g.cpp"
 #include "winrt/Windows.Globalization.h"
+#include <wil\resource.h>
 
 const wchar_t c_languageQualifierName[] = L"Language";
 
 namespace winrt::Microsoft::Windows::ApplicationModel::Resources::implementation
 {
+
+ResourceContext::~ResourceContext()
+{
+    winrt::slim_lock_guard guard {m_lock};
+    MrmDestroyResourceContext(m_resourceContext);
+}
+
+
+winrt::Windows::Foundation::Collections::IMap<hstring, hstring> ResourceContext::QualifierValues()
+{
+    m_lock.lock_shared();
+
+    auto cleanupOnFailure = wil::scope_exit([&] { m_lock.unlock_shared(); });
+
+    if (!IsInitialized())
+    {
+        m_lock.unlock_shared();
+        m_lock.lock();
+
+        InitializeQualifierValueMap();
+
+        m_lock.unlock();
+        m_lock.lock_shared();
+    }
+
+    return m_qualifierValueMap;
+}
+
+void ResourceContext::Apply()
+{
+    winrt::slim_lock_guard guard {m_lock};
+
+    if (m_resourceContext == nullptr)
+    {
+        // Resource not handled by MRT. Nothing to apply.
+        return;
+    }
+
+    InitializeQualifierValueMap();
+
+    for (auto const& eachValue : m_qualifierValueMap)
+    {
+        if (!eachValue.Value().empty())
+        {
+            winrt::check_hresult(MrmSetQualifier(m_resourceContext, eachValue.Key().c_str(), eachValue.Value().c_str()));
+        }
+    }
+}
+
+MrmContextHandle ResourceContext::GetContextHandle()
+{
+    winrt::slim_shared_lock_guard guard {m_lock};
```

Comment:
>winrt::slim_shared_lock_guard guard {m_lock}; [](http://example.com/codeflow?start=4&length=45)

is this needed?

Reply:
Because there's a possibility the dtor get called on a separate thread.

Reply:
can dtor be called when you still have a reference?

Reply:
Can the app obtains a reference and then call the dtor? Sure, and there's nothing that can be done about that, at least not in the PR.

The caller gets a MrmContextHandle which is defined be some macros. I'm not familiar with how the type is defined but it doesn't seem to be ref counted, so whenever the dtor is called, the handle will be released and all references handed out will become dangling.

Introducing some sort of reference counting or other mechanism to cope with this is beyond the scope of this fix.

At least now we garantee that the dtor won't race with GetContextHandle.

Reply:
i think winrt object is already refcounted, and you don't have to worry about someone else calling dtor when you still hold a ref.

Reply:
Cool, thanks for confirming.